libobs/obs-source.c

```diff
@@ -860,11 +860,15 @@ uint32_t obs_get_source_output_flags(const char *id)
 
 static void obs_source_deferred_update(obs_source_t *source)
 {
-	if (source->context.data && source->info.update)
-		source->info.update(source->context.data,
-				    source->context.settings);
-
-	source->defer_update = false;
+	if (source->context.data && source->info.update) {
+		uint32_t count;
+		do {
+			count = source->defer_update_count;
+			source->info.update(source->context.data,
+					    source->context.settings);
+		} while (!atomic_compare_exchange_strong(
+			&source->defer_update_count, &count, 0));
```

Comment:
Alternatively if we are afraid of blocking for too long, we can only do one update instead of the loop and let the next update be handled by the next video tick.